src/Security/PasswordExpirationMiddleware.php

```diff
@@ -0,0 +1,328 @@
+            if ($response = $this->handleExpiredPassword($request)) {
+                return $response;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Check if we have a redirect to a password change form registered
+     * and redirect there if possible.
+     * Otherwise, deauthenticate the user by resetting it for this request,
+     * since we should treat ones with expired passwords as unauthorised.
+     *
+     * @param HTTPRequest $request
+     *
+     * @return HTTPResponse|null
+     */
+    protected function handleExpiredPassword(HTTPRequest $request): ?HTTPResponse
+    {
+        $session = $request->getSession();
+
+        $sessionRedirectUrl = $session->get(static::SESSION_KEY_REDIRECT);
+        $defaultRedirectUrl = static::config()->get('default_redirect');
+
+        if ($sessionRedirectUrl || $defaultRedirectUrl) {
+            $redirectUrl = $this->absoluteUrl($sessionRedirectUrl ?? $defaultRedirectUrl);
+        } else {
+            $redirectUrl = null;
+        }
+
+        if (!$session || !$redirectUrl) {
+            Security::setCurrentUser(null);
+            return null;
+        }
+
+        $whitelist = [];
+
+        $configWhitelist = static::config()->get('whitelisted_urls');
+        if (is_array($configWhitelist)) {
+            $whitelist = array_merge($whitelist, $configWhitelist);
+        }
+
+        $sessionWhitelist = $session->get(static::SESSION_KEY_WHITELIST);
+        if (is_array($sessionWhitelist)) {
+            $whitelist = array_merge($whitelist, $sessionWhitelist);
+        }
+
+        $currentUrl = $this->absoluteUrl($request->getURL(true));
+
+        $allowedUrls = array_merge([
+            $redirectUrl
+        ], array_map(
+            function ($url) {
+                return $this->absoluteUrl($url);
+            },
+            $whitelist
+        ));
+
+        if (in_array($currentUrl, $allowedUrls, true)) {
+            return null;
+        }
+
+        $allowedStartswith = static::config()->get('whitelisted_url_startswith');
+        if (is_array($allowedStartswith)) {
+            foreach ($allowedStartswith as $pattern) {
+                $startswith = $this->absoluteUrl($pattern);
+
+                if (strncmp($currentUrl, $startswith, strlen($startswith)) === 0) {
+                    return null;
+                }
+            }
+        }
+
+        return $this->redirectOrForbid($request, $redirectUrl);
+    }
+
+    /**
+     * Builds an absolute URL for the given path, adds base url
+     * if the path configured as absolute
+     *
+     * @param string $url
+     *
+     * @return string
+     */
+    protected static function absoluteUrl($url): string
+    {
+        if (substr($url, 0, 1) === '/') {
+            // add BASE_URL explicitly if not absolute
+            $url = Controller::join_links(Director::absoluteBaseURL(), $url);
+        } else {
+            $url = Director::absoluteURL($url);
```

Comment:
Actually, why can't you just use `Director::absoluteURL()` for this? It should handle the "is root relative" logic internally

Reply:
> why can't you just use Director::absoluteURL()

Because `absoluteURL` would return incorrect results for `SS_BASE_URL` which has non-root base.
E.g.
```php
Environment::set('SS_BASE_URL', 'http://localhost/non-root/');

$a = Director::absoluteURL('/Secret/');

$a === 'http://localhost/Secret/'; // we want http://localhost/non-root/Secret/
```

Reply:
I realise you'd been discussing this with @tractorcow recently, but that sounds like a bug to me and could be addressed in a separate issue

Reply:
That's not really a bug per se, but rather `Director::absoluteURL` doesn't cover the particular use case we have right here.

We already have a separate issue for discussing that: https://github.com/silverstripe/silverstripe-framework/issues/9111

Reply:
> Strict typing wise, this method can return false

yeah, I was tricked by the Director::absoluteURL docbloc :smile: